### PR TITLE
PM UI:  enable remoting of install/uninstall package functionality

### PR DESCRIPTION
**Pull request by dtivel**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9526 and https://github.com/NuGet/Home/issues/9553
Regression: No  

## Fix

This change enables install package and uninstall package functionality in PM UI to be remoted.

*  I'm experiencing build failures locally after my last rebase.  Something about test projects not targeting .NET 5.
*  Update package was broken before this change and remains broken.
*  The `ISolutionManager` predicament needs to be resolved ASAP.  I implemented a temporary workaround in this PR to enable refreshing PM UI after a package operation.

## Testing/Validation

Tests Added: Yes, but not for remote scenarios.
Reason for not adding tests:  
Validation:

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs [hunk 1]
```diff
@@ -73,48 +103,384 @@ public async ValueTask<IReadOnlyCollection<PackageReference>> GetInstalledPackag
...
+
+            NuGetProject project = await GetNuGetProjectMatchingProjectIdAsync(projectId);
 
             bool success = project.TryGetMetadata(key, out object value);
+
             return (success, value);
         }
 
-        public async ValueTask<bool> IsProjectUpgradeableAsync(string projectGuid, CancellationToken cancellationToken)
+        public async ValueTask BeginOperationAsync()
```
Comment (sbanni) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs]: What would happen if this was called by two different projects at the same time?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/NuGetUI.cs [hunk 2]
```diff
@@ -259,6 +248,13 @@ public void OnActionsExecuted(IEnumerable<ResolvedAction> actions)
             UIContext.SolutionManager.OnActionsExecuted(actions);
         }
 
+        internal event EventHandler<IReadOnlyCollection<string>> ActionsExecuted;
+
+        internal void OnActionsExecuted(IReadOnlyCollection<string> projectIds)
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/NuGetUI.cs:253]: uncalled member. still to be done, i'm assuming.

File: src/NuGet.Clients/NuGetConsole.Host.PowerShell/NuGetHostUserInterface.cs [hunk 3]
```diff
@@ -22,7 +22,7 @@ internal class NuGetHostUserInterface : PSHostUserInterface, IHostUISupportsMult
         public const ConsoleColor NoColor = (ConsoleColor)(-1);
         private const int VkCodeReturn = 13;
         private const int VkCodeBackspace = 8;
-        private static Color[] _consoleColors;
+        private static Color[] ConsoleColors;
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGetConsole.Host.PowerShell/NuGetHostUserInterface.cs:25]: this file only has formatting improvements/etc. worth including with no real changes?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj [hunk 4]
```diff
@@ -233,6 +233,10 @@
       <Project>{32a23995-14c7-483b-98c3-0ae4185373ea}</Project>
       <Name>NuGet.Credentials</Name>
     </ProjectReference>
+    <ProjectReference Include="..\..\NuGet.Core\NuGet.Resolver\NuGet.Resolver.csproj">
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj:236]: no code change. why add?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs [hunk 5]
```diff
@@ -660,6 +909,12 @@ List<TelemetryEvent> ToTelemetryPackageList(List<Tuple<string, string>> packages
                     .OrderBy(name => name, StringComparer.OrdinalIgnoreCase)
                     .ToList();
 
+                await Task.WhenAll(tasks);
```
Comment (sbanni) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: >await Task.WhenAll(tasks); [](start = 16, length = 26)

This is happening already in the assignment of 'string[] projectNames'

**Review by sbanni — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/INuGetUIContext.cs [hunk 6]
```diff
@@ -37,5 +39,7 @@ public interface INuGetUIContext : IDisposable
         Task<bool> IsNuGetProjectUpgradeableAsync(IProjectContextInfo project, CancellationToken cancellationToken);
 
         Task<IModalProgressDialogSession> StartModalProgressDialogAsync(string caption, ProgressDialogData initialData, INuGetUI uiService);
+
+        void FireProjectActionsExecuted(IReadOnlyCollection<ProjectAction> projectActions);
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/INuGetUIContext.cs]: .NET DESIGN GUIDELINES say to stay away from Fire or Trigger, in favor for Raise.
[https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event)

**Review by rrelyea — APPROVED** [review]
:shipit:

**Pull request merged**